dialogue-core/src/main/java/com/palantir/dialogue/core/PreferLowestUtilization.java

```diff
@@ -0,0 +1,118 @@
+    private final ImmutableList<LimitedChannel> channels;
+
+    /** Each AtomicInteger stores the number of active requests that a channel is currently serving. */
+    private final LoadingCache<LimitedChannel, AtomicInteger> activeRequestsPerChannel =
+            Caffeine.newBuilder().maximumSize(1000).build(upstream -> new AtomicInteger());
+
+    PreferLowestUtilization(ImmutableList<LimitedChannel> channels, Random random) {
+        Preconditions.checkState(channels.size() >= 2, "At least two channels required");
+        this.channels = channels;
+        this.random = random;
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+
+        // we accumulate everything right now (which is quite expensive), but it allows us to move on to the
+        // next-best channel if our preferred one refuses
+        ListMultimap<Integer, LimitedChannel> channelsByUtilization = sortChannelsByLowestUtilization();
+
+        for (Integer activeCount : channelsByUtilization.keySet()) {
+            // multiple channels might have the same number of inflight requests
+            for (LimitedChannel channel : channelsByUtilization.get(activeCount)) {
+
+                AtomicInteger atomicInteger = activeRequestsPerChannel.get(channel);
+                atomicInteger.incrementAndGet();
+
+                Optional<ListenableFuture<Response>> maybeResponse = channel.maybeExecute(endpoint, request);
+
+                if (maybeResponse.isPresent()) {
+                    ListenableFuture<Response> response = maybeResponse.get();
+                    response.addListener(atomicInteger::decrementAndGet, MoreExecutors.directExecutor());
+                    return Optional.of(response);
+                } else {
+                    // we quickly undo the atomicInteger thing we optimistically incremented earlier
+                    atomicInteger.decrementAndGet();
+                }
+            }
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Every channel refused", SafeArg.of("channels", channelsByUtilization));
+        }
+        return Optional.empty();
+    }
+
+    private ListMultimap<Integer, LimitedChannel> sortChannelsByLowestUtilization() {
+        ListMultimap<Integer, LimitedChannel> channelsByUtilization =
+                MultimapBuilder.treeKeys().arrayListValues().build();
+
+        // the shuffle ensures that when nodes have the same utilization, we won't always pick the same one
+        for (LimitedChannel channel : shuffleImmutableList(channels)) {
+            int activeRequests = activeRequestsPerChannel.get(channel).get();
+            channelsByUtilization.put(activeRequests, channel);
+        }
+
+        return channelsByUtilization;
+    }
+
+    /** Returns a new shuffled list, without mutating the input list (which may be immutable). */
+    private <T> List<T> shuffleImmutableList(ImmutableList<T> sourceList) {
```

Comment:
Not sure we need to go so far as sorting lists here, but there's a quick optimization in which we sort the result of the inner for lop in `maybeExecute` where we can no-op the simple case that sourceList has fewer than two elements.

Reply:
Longer term this object knows when each channel receives and completes a request, so it can use that to maintain an internal data structure that doesn't require building b-trees for each request.